### DISCUSSION TITLE
Can I send heap allocated data structures with the UserDataType API?

### DISCUSSION THREAD

**Discussion by Janekdererste** [rsmpi/rsmpi#147] [Q&A]
Hi, 

I am a MPI beginner trying to wrap my head around the various option this framework has. In my application I want to send complex data types, including `Vec`s of structs, where I don't know the number of things I have to send beforehand. Possibly, I will have to send `Strings` too. 

I started looking into the [complex datatypes](https://github.com/rsmpi/rsmpi/blob/main/examples/complex_datatype.rs) example on how to pass structured data to the MPI implementation. As far as I have understood now, we pass a pointer to the address of the struct and then via the `Equivalence` trait tell the implementation what the following bytes mean. 

Is it possible to use this API with heap allocated data structures as well? If not, what is a good way to handle this type of data?

**Comment by jedbrown** [comment]
Hi :wave: you can indeed work with heap-allocated data, but the data that you're sending should be plain data, not pointers/boxes/etc. Composite data structures can be serialized and sent in that form.

**Comment by Janekdererste** [comment]
Thanks @jedbrown for your suggestion. I can't figure out from the available examples how I would create a message from a data structure with a pointer (e.g. a Vec).

So far I decided to just serialize my data structures using protobuf and pass the serialized message to resmpi `send`.

**Reply by hppritcha** [reply]
@Janekdererste you have hit upon an important topic.  A fundamental premise of the MPI datatype approach is that the programmer knows in detail the layout of the data they want to send in terms that can be expressed via the methods described in the MPI standard's datatype chapter (which moves around a bit depending on the version of the standard you are using).   This was fine for languages like Fortran and c, but doesn't work that well for programming languages that make extensive use of opaque objects, etc.  

What @jedbrown suggests is probably the best that can be done for now.

**Reply by Janekdererste** [reply]
I see. This makes sense. Thank you very much for your response.

**Comment by lquenti** [comment]
To further elaborate: One can slice the vector completely (i.e. `my_vec[..]) which makes it a trivial datatype. This datatype can then be put into a helper struct, which can derive the `Equivalence` typeclass (given the crate feature is enabled)

(sorry for the necrobump)
